### ISSUE TITLE
Wildcards Questions

### ISSUE THREAD

**Issue by Guo-Yujia** [nus-cs2030/2021-s2#365]
# Summary
I didn't have a good command of the bounded wildcards part and I couldn't fully understand when we should use upper bounded and lower bounded wildcards. Could someone explain it again please?

## Screenshots (if any): 
Here are some examples from lecture notes.
P13
![image](https://user-images.githubusercontent.com/77218700/116066049-36e46e00-a6ba-11eb-9bcb-de162334487b.png)

P14
![image](https://user-images.githubusercontent.com/77218700/116066096-44015d00-a6ba-11eb-8a2e-0f2725680a97.png)

P15
![image](https://user-images.githubusercontent.com/77218700/116066191-5bd8e100-a6ba-11eb-865f-531649778154.png)

**Comment by brendancjz** [comment]
Hi! Here's a link that helped me to better understand this concept: https://stackoverflow.com/questions/2723397/what-is-pecs-producer-extends-consumer-super

In essence, we use <? extends T> for Producers. An example will be to imagine we have an arrayList of T objects. We want to use the items in the arrayList. So we can get items from this list and use the methods of this item object. So, this arrayList is a producer because it is producing us the items. And we can call item.foo() because we know that .foo() is implemented in T item. Now, we want to make this arrayList to be more flexible but still does the job. So, instead of writing arrayList<T>, we write arrayList<? extends T>. This is very possible because we know that all items in this arrayList can still call the .foo() as these items are either T or T subclasses. So, using this upper bound wildcard to achieve this flexibility.

For <? super T>, we use it for Consumers. To be honest, I am not as familiar with this concept. But from looking at the lecture slides, Predicate and Comparator and Function uses <? super T> takes in the super classes of T so that when T is consumed by them, the T item can call its super classes methods no problem. I'm stilling trying to understand this haha hope it helps!

**Comment by clementyee303** [comment]
> 
> 
> Hi! Here's a link that helped me to better understand this concept: https://stackoverflow.com/questions/2723397/what-is-pecs-producer-extends-consumer-super
> 
> In essence, we use <? extends T> for Producers. An example will be to imagine we have an arrayList of T objects. We want to use the items in the arrayList. So we can get items from this list and use the methods of this item object. So, this arrayList is a producer because it is producing us the items. And we can call item.foo() because we know that .foo() is implemented in T item. Now, we want to make this arrayList to be more flexible but still does the job. So, instead of writing arrayList, we write arrayList<? extends T>. This is very possible because we know that all items in this arrayList can still call the .foo() as these items are either T or T subclasses. So, using this upper bound wildcard to achieve this flexibility.
> 
> For <? super T>, we use it for Consumers. To be honest, I am not as familiar with this concept. But from looking at the lecture slides, Predicate and Comparator and Function uses <? super T> takes in the super classes of T so that when T is consumed by them, the T item can call its super classes methods no problem. I'm stilling trying to understand this haha hope it helps!

The reason `? super T` is used is to allow for wider types to be compilable / accepted as an input parameter. 

E.g. 

Optional<Integer> optional = Optional.of(1);
Function<Number, String> mapper = i -> String.valueOf(i);
optional.map(mapper::apply); // no compile error

This is because  Optional's map function takes in (Function<? super T, ? extends U> mapper)
Hence, for an Optional<**Integer**>, Function<**Number**, ? extends U> can be accepted. (Integer <: Number)

However, what if Optional's map function only takes in (Function<T, ? extends U> mapper) ?
Now, for an Optional<**Integer**>, only Function<**Integer**, ? extends U> can be accepted. 
This means:
Optional<Integer> optional = Optional.of(1);
Function<Number, String> mapper = i -> String.valueOf(i);
optional.map(mapper::apply); // compile error

**Comment by wglee99** [comment]
Hello, just to add on to what was mentioned by brendan, although PECS is a good acronym to remember wildcards with, I found the last recitation (or one of the last few) where Prof Henry went through the finer details of upper and lower bounded wildcards pretty helpful in trying to understand this.

Basically, for Consumers, you're taking in something. In this case, you're taking in an argument/input of type T. 

But what if the client doesn't know that this function takes in an input of type T? So you do a <? super T> which allows for the inputting of T and its super classes, since if it can take in a superclass of type T, then it can definitely take in type T.

Similarly, for Producer, you're producing an output, in this case of type R.

Here you do a <? extends R> which allows for the output of R and its sub classes, since if you're expecting output of type R then it's still ok if the output is of a subtype of R.

As has been mentioned, this is a widening type conversion, which is essentially relaxing the input type and output type to allow for a wider range of types.

**Comment by Guo-Yujia** [comment]
That's much more clear, thanks for helping! So just to double check for BiFunction<T, U, R>, it is BiFunction<? super T, ? super U, ? extends R>, isn't it? Because it ensures that super classes of T and U can be input and the return value can be subclasses of R.

**Comment by Buwoo** [comment]
> That's much more clear, thanks for helping! So just to double check for BiFunction<T, U, R>, it is BiFunction<? super T, ? super U, ? extends R>, isn't it? Because it ensures that super classes of T and U can be input and the return value can be subclasses of R.

Yes, I believe its correct to say BiFunction<? super T, ? super U, ? extends R>. In terms of PECS, objects of type T and U are being consumed by the BiFunction, therefore we use super. Since objects of type R are being produced, we use extends. 
And yes, you're correct in saying that this "ensures that super classes of T and U can be input and the return value can be subclasses of R"!

**Comment by JWulaXia** [comment]
BiFunction<T, U, R> -->  BiFunction<? super T, ? super U, ? extends R>
i think it is now much more clearer
